### ISSUE TITLE
Publicly Accessible UploadTask

### ISSUE THREAD

**Issue by tcldr** [segmentio/analytics-swift#47]
**Is your feature request related to a problem? Please describe.**
Currently, on iOS, the only way to run an upload as a well behaved background task seems to be to opt-in to lifecycle events. That's because the current implementation seems to rely on an internal UploadTaskInfo in the SegmentDestination which is only accessible to internal clients.

AFAICT this means there is no simple way to create your own background task to upload any outstanding events.

This would be especially useful in low/no network situations. Currently, even if opted-in to iOS lifecycle events, the background task that is created will fail if the network is unavailable.

**Describe the solution you'd like**
An ideal solution would be one with flexibility and adherence to platform recommendations.

For example, on iOS it is possible to opt-in to 'background modes' one of which is called 'background processing'. This mode allows a developer to use he system's Background Task Scheduler to do additional processing at some point when the battery and network conditions are favourable. The system scheduler enforces adherence to platform recommendations (i.e cancelling when asked, and letting the system know when it's complete) by docking an app's time on the scheduler for non-compliance.

With that in mind, I suggest a task API similar to the one for URLSession but with an additional callback for completion with some kind of TaskState to inform of success, cancellation or failure:

```
struct SegmentUploadTask {
    func start(_ completion: @escaping (TaskState) -> Void)
    func cancel()
}
```

**Describe alternatives you've considered**

**Additional context**

**Comment by bsneed** [comment]
Hi @tcldr, I'm not sure I understand what you're asking for here.

Background segment uploads are handled via a SegmentDestination extension in iOSLifecycleMonitor.  I don't think we'd want to push that out to the customer to implement on their own.

**Comment by tcldr** [comment]
So, afaict at the moment it uses beginBackgroundTask(expirationHandler:) to attempt a last push to Segment’s server as the app resigns active. And as far as I know this particular task type won’t postpone or rerun if there is no network.

So if that fails I’d like to be able to attempt an upload when the network finally is available.

Going back to my favourite scenario, first launch after installation, say I have a customer who launches the app for the first time on the Tube (no network). And they’re not impressed. So they exit the app and forget about it. Maybe they delete it a week later when they’re de-cluttering their home screen.

Here I’ve got a bunch of stuff that requires the network. Attribution, Segment settings, and event upload.

And as the user is in a no network situation when they exit the app, the background task won’t complete. So my understanding is I’ll never get the events.

I could setup a background processing task myself (different from the UIApplication API - https://developer.apple.com/documentation/backgroundtasks/bgprocessingtask) but really I need a completion hook on the flush api so I can exit once I know it’s done.

It means I’m much more likely to get representative data from those users that launch only once.

**Comment by bsneed** [comment]
Ahh... I think I get it now.

Would a version of flush() with an optional completion handler suffice for what you wanna do?  And if so, would you be willing to do a utility plugin we could include in the examples for others to use/copy/modify?

**Comment by bsneed** [comment]
Actually, n/m, that idea I had won't work.  It's not structured like that.  Flush ends up calling flush on every destination and I can't tell when all of them would be complete.  Even the segment flush issues multiple uploads depending on what's waiting to go out.  Will continue to think on this.

**Comment by tcldr** [comment]
A completion handler would be good, but I’d forgotten that cancellation is important too. If the system needs resources elsewhere it gives background tasks an opportunity to exit cleanly. I understand the system applies penalties to apps with tasks that don’t exit promptly when cancellation is requested by the system. e.g. no longer given time on the background scheduler.

That’s why I originally suggested the task style API above as it seemed the most idiomatic for these purposes.

**Comment by bsneed** [comment]
@tcldr would you have time at some point to jump on a zoom call with me and discuss this further?

**Comment by tcldr** [comment]
Yeah, we can do that.

Theoretically the operation *should* be quick if the device has decent network – it’s not downloading 100s of images or anything.  On the other hand, i don’t know what the plug-ins that people build will do.

But I would think a completion handler would be bare minimum to make use of Apple’s BG tasks API as you need to mark your task as ‘done’ in order for the scheduler to move on.

There’s more on BGTaskScheduler here: https://developer.apple.com/documentation/backgroundtasks/bgtaskscheduler
And the session it was introduced here: https://developer.apple.com/wwdc19/707

**Comment by bsneed** [comment]
Doesn't seem like we're gonna be able to do this anytime soon unfortunately.
